File: app/src/main/java/win/deri/ca/qqfucker/Hook.java
package win.deri.ca.qqfucker;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;

import java.lang.reflect.Method;
import java.util.Arrays;

import de.robv.android.xposed.IXposedHookLoadPackage;
import de.robv.android.xposed.XC_MethodHook;
import de.robv.android.xposed.XC_MethodReplacement;
import de.robv.android.xposed.XposedBridge;
import de.robv.android.xposed.XposedHelpers;
import de.robv.android.xposed.callbacks.XC_LoadPackage.LoadPackageParam;

public class Hook implements IXposedHookLoadPackage {
    private interface FilterFunction {
        boolean shouldHook(Method method);
    }

    private interface CallbackFunction {
        Object cb(XC_MethodHook.MethodHookParam param);
    }

    private void findAndHook(LoadPackageParam lpp, String className, FilterFunction filter, CallbackFunction callback) {
        final Class Cls = XposedHelpers.findClass(className, lpp.classLoader);
        if (Cls == null) return;
        Arrays.stream(Cls.getDeclaredMethods())
                .filter(filter::shouldHook)
                .forEach((method) -> hookMethod(className, method, callback));
    }

    private void hookMethod(String className, Method method, CallbackFunction callback) {
        XposedBridge.hookMethod(method, new XC_MethodReplacement() {
            @Override
            protected Object replaceHookedMethod(MethodHookParam param) {
                XposedBridge.log("QQFucker: fucked " + className + "." + method.getName());
                return callback.cb(param);
            }
        });
    }

    private void hook(LoadPackageParam lpp, String className, String methodName) {
        findAndHook(lpp, className, (method) -> method.getName().equals(methodName), (XC_MethodHook.MethodHookParam param) -> null);
    }

    private void hook(LoadPackageParam lpp, String className, String methodName, Integer methodCount, CallbackFunction callback) {
        findAndHook(lpp, className, (method) -> method.getName().equals(methodName) && method.getParameterCount() == methodCount, callback);
    }

    private CallbackFunction openWithBrowser = (XC_MethodHook.MethodHookParam param) -> {
        Context activity = (Context) param.args[0];
        String uri = (String) param.args[1];
        Intent intent = new Intent(Intent.ACTION_VIEW)
                .setData(Uri.parse(uri))
                .putExtra("exlink", true)
                .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        activity.startActivity(intent);
        return intent;
    };

    public void handleLoadPackage(final LoadPackageParam lpp) {
        if ("com.tencent.qqlite".equals(lpp.packageName)) {
            hook(lpp, "com.tencent.mobileqq.mini.sdk.MiniAppLauncher", "startMiniApp");
            hook(lpp, "jvo", "F");
            hook(lpp, "jxl", "onClick");
            hook(lpp, "cooperation.qwallet.plugin.QWalletHelper", "launchQWalletAct");
            hook(lpp, "kri", "onClick");
            hook(lpp, "krk", "onClick");
            hook(lpp, "krm", "onClick");
            hook(lpp, "krg", "onClick");
            hook(lpp, "cooperation.qzone.QzonePluginProxyActivity", "a");
            hook(lpp, "oki", "a", 8, openWithBrowser);
        }
    }
}

File: app/src/main/java/winderi/ca/qqfucker/Hook.kt
package winderi.ca.qqfucker

import android.content.Context
import android.content.Intent
import android.net.Uri
import de.robv.android.xposed.*
import de.robv.android.xposed.XC_MethodHook.MethodHookParam
import de.robv.android.xposed.callbacks.XC_LoadPackage.LoadPackageParam
import java.lang.reflect.Method

class Hook : IXposedHookLoadPackage {
    private fun filterAndHook(
        classLoader: ClassLoader,
        className: String,
        filter: (Method?) -> Boolean,
        callback: (MethodHookParam?) -> Any? = { null }
    ) {
        XposedHelpers.findClass(className, classLoader).declaredMethods
            .filter { filter(it) }
            .forEach {
                XposedBridge.hookMethod(it, object : XC_MethodReplacement() {
                    override fun replaceHookedMethod(param: MethodHookParam): Any? {
                        XposedBridge.log("QQFucker: fucked $className.${it.name}")
                        return callback(param)
                    }
                })
            }
    }

    override fun handleLoadPackage(lpp: LoadPackageParam) {
        if (lpp.packageName != "com.tencent.qqlite") {
            return
        }
        val classLoader = lpp.classLoader
        mapOf(
            "com.tencent.mobileqq.mini.sdk.MiniAppLauncher" to "startMiniApp",
            "cooperation.qwallet.plugin.QWalletHelper" to "launchQWalletAct",
            "cooperation.qzone.QzonePluginProxyActivity" to "a",
            "jvo" to "F",
            "jxl" to "onClick",
            "kri" to "onClick",
            "krk" to "onClick",
            "krm" to "onClick",
            "krg" to "onClick",
            "com.tencent.mobileqq.app.upgrade.UpgradeController" to "f"
        ).forEach { (key, value) -> filterAndHook(classLoader, key, { it?.name == value }) }
        filterAndHook(classLoader, "qhh", { it?.name == "a" && it.parameterCount == 8 }) {
            it?.let {
                (it.args[0] as Context).startActivity(
                    Intent(Intent.ACTION_VIEW)
                        .setData(Uri.parse(it.args[1] as String))
                        .putExtra("exlink", true)
                        .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                )
            }
        }
        filterAndHook(classLoader, "rtp", { it?.name == "a" && it.parameterCount == 4 }) {
            it?.let {
                if (it.args[0].javaClass.name != "rts") {
                    XposedBridge.invokeOriginalMethod(it.method, it.thisObject, it.args)
                }
            }
        }
        XposedHelpers.findAndHookMethod(
            "rkk", classLoader, "b", Int::class.java, object : XC_MethodHook() {
                override fun beforeHookedMethod(param: MethodHookParam?) {
                    param?.let {
                        if (it.args[0] == -5008) {
                            it.args[0] = -1000
                        }
                    }
                }
            }
        )
    }
}